cpmpy/expressions/globalconstraints.py

```diff
@@ -672,83 +672,162 @@ def __repr__(self):
+        raise Exception
+
+    def _task_decomposition(self):
         """
-            Time-resource decomposition from:
+         Task-based decomposition of the cumulative constraint.
             Schutt, Andreas, et al. "Why cumulative decomposition is not as bad as it sounds."
             International Conference on Principles and Practice of Constraint Programming. Springer, Berlin, Heidelberg, 2009.
         """
+        start, duration, end, demand, capacity = self.args
 
-        arr_args = (cpm_array(arg) if is_any_list(arg) else arg for arg in self.args)
-        start, duration, end, demand, capacity = arr_args
+        cons = [d >= 0 for d in duration]  # enforce non-negative durations
+        cons += [h >= 0 for h in demand]  # enforce non-negative demand
 
-        cons = []
+        # set duration of tasks, only if end is user-provided
+        if end is None:
+            end = [start[i] + duration[i] for i in range(len(start))]
+        else:
+            cons += [start[i] + duration[i] == end[i] for i in range(len(start))]
+
+        # demand doesn't exceed capacity
+        for i in range(len(start)):
+            demand_at_i = demand[i]
+            for j in range(len(start)):
+                if i != j:
+                    demand_at_i += demand[j] * ((start[j] <= start[i]) & (end[j] > start[i]))
 
-        # set duration of tasks
-        for t in range(len(start)):
-            cons += [start[t] + duration[t] == end[t]]
+            cons += [demand_at_i <= capacity]
+
+        return cons, []
+
+    def _time_decomposition(self):
+        """
+         Time-resource decomposition of the cumulative constraint.
+            Schutt, Andreas, et al. "Why cumulative decomposition is not as bad as it sounds."
+            International Conference on Principles and Practice of Constraint Programming. Springer, Berlin, Heidelberg, 2009.
+        """
+        start, duration, end, demand, capacity = self.args
+
+        cons = [d >= 0 for d in duration] # enforce non-negative durations
+        cons += [h >= 0 for h in demand] # enforce non-negative demand
+
+        # set duration of tasks, only if end is user-provided
+        if end is None:
+            end = [start[i] + duration[i] for i in range(len(start))]
+        else:
+            cons += [start[i] + duration[i] == end[i] for i in range(len(start))]
 
         # demand doesn't exceed capacity
-        lb, ub = min(get_bounds(start)[0]), max(get_bounds(end)[1])
+        lbs, ubs = get_bounds(start)
+        lb, ub = min(lbs), max(ubs)
         for t in range(lb,ub+1):
             demand_at_t = 0
             for job in range(len(start)):
-                if is_num(demand):
-                    demand_at_t += demand * ((start[job] <= t) & (t < end[job]))
-                else:
-                    demand_at_t += demand[job] * ((start[job] <= t) & (t < end[job]))
+                demand_at_t += demand[job] * ((start[job] <= t) & (end[job] > t))
```

Comment:
like above, please do this as a list instead of the less readable accumulating sum, so that it shows cons.add( cp.sum(demands_at_t) <= capacity )

Or you could actually also make a helper function demands_at(t) and then
for t in ...:
   cons.add( cp.sum(demands_at(t)) <= capacity )
   
 not sure what is more readable